accepted/future-releases/nnbd/feature-specification.md

```diff
@@ -616,6 +609,245 @@ It is a warning to use a null aware operator (`?.`, `?[]`, `?..`, `??`, `??=`, o
+}
+```
+
+Local variables with no explicitly written type but with an initializer are
+given an inferred declared type equal to the type of their initializer.  In the
+case that the type of the initializer is a promoted type variable `X & T`, the
+inferred type of the variable shall be `X`.  However, such a variable shall be
+treated as immediately promoted to `X & T`.
+
+In the null safety release, local variables with no explicitly written type and
+no initializer are subject to further promotion based inference.  The variables
+under consideration here are those declared as `var x;`, `final x;`, `late var
+x;`, or `late final x;`.  We refer to such variables as **untyped variables**.
+
+An **initializing assignment** is an assignment to an **untyped variable** when
+that variable is in the **definitely unassigned** state.  Such an assignment on
+a path shall cause the variable in question to be treated by the flow analysis
+on that path as if it were declared with the assigned expression as its
+initializing expression, up until the next join point in the program.  We refer
+to this inferred type as the **initialization inferred type** of the variable.
+The **initialization inferred type** of the variable is always treated as a
+**type of interest** for the purposes of promotion (including its non-nullable
+version if applicable).
+
+```dart
+void test() {
+  var x; // x is an untyped variable
+  x = 3; // x is given the initialization inferred type of `int`
+}
+```
+
+At a join point in the program, it is an error if an **untyped variable** has
+been given an **initialization inferred type** on one of the reachable paths to
+the join point and not on another reachable path, unless the variable is
+declared as `late`.  Note that it is valid for a variable to have no
+**initialization inferred type** on a path into a join point which is known to
+be unreachable.
+
+```dart
+void test(bool b) {
+  var x; // x is an untyped variable
+  if (b) {
+    x = 3;
+  } // Error, x has not been given an inferred type on all paths
+
+  late var y; // y is a late untyped variable
+  if (b) {
+    y = 3;
+  } // Ok, y has inferred type `int`
+
+  var z;
+  if (b) {
+     z = 3;
+  } else {
+    return;
+  }  // No error: z has been given a type on all reachable paths into the join
+}
+```
+
+At a join point in the program, if an **untyped variable** has been given an
+**initialization inferred type** on more than one of the reachable paths to the
+join point, the variable is treated after the join point as having the
+**initialization inferred type** equal to the upper bound of the
+**initialization inferred types** from the reachable paths into the join point.
+The **initialization inferred type** so computed is treated as a **type of
+interest** for the variable in question (including its non-nullable version if
+applicable).  The promoted type chains for the variable on the respective paths
+are intersected as usual.
```

Comment:
I can't read from this what the following program does:
```dart
int? v = ...;
var x;
if (test) {
  x = v;  // initializing assingment, type int?.
  if (x == null) x = 42;
  // x has type int.
} else {
  x = 0;
}
// What is type of x here?

That is, *what are* the chains in question. Is the initialization inferred type part of the chain or not?

Reply:
As specified here, the IIT is not part of the chain, and the type of `x` after the final join in your example will be `int?`.  It is probably reasonable to add the IIT to the chain at some point, albeit a bit weird (usually we have the elements of the chain are proper subtypes of the declared type).  Note that this won't fix all of the examples that you were interested in.  The final IIT for this example will still be `int?`, it's just that it will have been promoted to `null` after the join.

Reply:
I have uploaded a patch specifying this, take a look.